docs/aws-config.md

```diff
@@ -149,6 +149,119 @@ Specify the following values in the example files:
+
+> This path assumes you have `kube2iam` already running in your Kubernetes cluster. If that is not the case, please install it first, following the docs here: https://github.com/jtblin/kube2iam
+
+It can be set up for Ark by creating a role that will have required permissions, and later by adding the permissions annotation on the ark deployment to define which role it should use internally.
+
+1. Create a Trust Policy document to allow the role being used for EC2 management & assume kube2iam role:
+
+    ```bash
+    cat > heptio-ark-role-trust-policy.json <<EOF
+    {
+        "Version": "2012-10-17",
+        "Statement": [
+            {
+                "Effect": "Allow",
+                "Principal": {
+                    "Service": "ec2.amazonaws.com"
+                },
+                "Action": "sts:AssumeRole"
+            },
+            {
+                "Effect": "Allow",
+                "Principal": {
+                    "AWS": "arn:aws:iam::<AWS_ACCOUNT_ID>:role/<ROLE_CREATED_WHEN_INITIALIZING_KUBE2IAM>"
+                },
+                "Action": "sts:AssumeRole"
+            }
+        ]
+    }
+    EOF
+    ```
+
+2. Create the IAM role:
+
+    ```bash
+    aws iam create-role --role-name heptio-ark-role --assume-role-policy-document file://./heptio-ark-role-trust-policy.json
+    ```
+
+3. Attach policies to give `heptio-ark-role` the necessary permissions:
+
+    ```bash
+    BUCKET=<YOUR_BUCKET>
+    cat > heptio-ark-role-policy.json <<EOF
+    {
+        "Version": "2012-10-17",
+        "Statement": [
+            {
+                "Effect": "Allow",
+                "Action": [
+                    "ec2:DescribeVolumes",
+                    "ec2:DescribeSnapshots",
+                    "ec2:CreateTags",
+                    "ec2:CreateVolume",
+                    "ec2:CreateSnapshot",
+                    "ec2:DeleteSnapshot"
+                ],
+                "Resource": "*"
+            },
+            {
+                "Effect": "Allow",
+                "Action": [
+                    "s3:GetObject",
+                    "s3:DeleteObject",
+                    "s3:PutObject"
+                ],
+                "Resource": [
+                    "arn:aws:s3:::${BUCKET}/*"
+                ]
+            },
+            {
+                "Effect": "Allow",
+                "Action": [
+                    "s3:ListBucket",
+                    "s3:AbortMultipartUpload",
+                    "s3:ListMultipartUploadParts"
+                ],
+                "Resource": [
+                    "arn:aws:s3:::${BUCKET}"
+                ]
+            }
+        ]
+    }
+    EOF
+    
+    aws iam put-role-policy \
+      --role-name heptio-ark-role \
+      --policy-name heptio-ark-role-policy \
+      --policy-document file://heptio-ark-role-policy.json
+    ```
+4. Update AWS_ACCOUNT_ID & HEPTIO_ARK_ROLE_NAME in the file `examples/common/10-deployment-kube2iam.yaml`:
```

Comment:
Rather than have a separate example file, would it be sufficient to have these instructions say that you need to add the appropriate annotation to `spec.template.metadata.annotations`? I'd like to avoid another variation of the deployment yaml if possible. WDYT?

Reply:
Hmm it is not just adding of the annotation, but also removal of the `AWS_SHARED_CREDENTIALS_FILE` env from the container, and removal of the `cloud-credentials` volume. I'm afraid that trying to just describe how those changes should be done on the existing `10-deployment.yaml` file is quite cumbersome.

Reply:
@domderen good point. Maybe it would be better to have `examples/common` for all the non-cloud specific bits, followed by `examples/aws`, `examples/azure`, `examples/gcp`, and `examples/minio` for the cloud-specific ones. And inside the aws directory, we could have 1 file using shared credentials, and another one for kube2iam. WDYT?

cc @skriss

Reply:
That separation makes sense to me. afaik kub2iam is only relevant to AWS, though.

Reply:
Makes sense, though it suddenly generates a lot more manifests, and I thought you wanted to avoid maintaining a lot of manifests? ;)

Reply:
Yeah.... but the more I thought about it, the more it makes sense to have multiple variations, so we don't have to tell people stuff like "take this AWS-based one and rename an env var so it'll work for Google"...

Reply:
Pushed a change where each cloud provider directory has it's own deployment file, and updated the docs where applicable.